### ISSUE TITLE
online disk grow/resizing with qemu/KVM

### ISSUE THREAD

**Issue by zitterbacke** [ganeti/ganeti#28]
Hi,

I like to reactivate an old goole-code-issue-258 I've found.

I found online grow more useful then adding a disk with --hotplug. The main reason is, that I can extend an existing guest filesystem without the usage of LVM (BTW: and even without the usage of partitions). Without LVM you have a 1:1 relation of a disk and a filesystem (in all layers). This makes a transfer of a disk between VMs (disk de-/attach functionality) more handy for me.

My workflow I use often is (pseudo code):

if disk_template in ['plain','drbd']:
  gnt-instance grow-disk ...
elif disk_template in ['blockdev']:
  grow-externally
  
if disk_template in ['plain','drbd','blockdev']:
  run_qemu_monitor_command: info block
  run_qemu_monitor_command: block_resize hotdisk-7d8a2ae9-pci-7 <any_size_smaller_then_current>G
elif disk_template in ['sharedfile','file']:
  run_qemu_monitor_command: info block
  run_qemu_monitor_command: block_resize hotdisk-7d8a2ae9-pci-7 <new_size_>G

if disk_template in ['sharedfile','file','blockdev']:
  gnt-cluster repair-disk-sizes

Sounds like a simple task to add this into Ganeti? Will have a look at it, hopefully not touching Haskell code ;-)

Thanks, Alf.

**Comment by kasimon** [comment]
A <pre>gnt-instance grow-disk --hotplug ...</pre> issuing the qemu commands above would indeed be a very useful addition to ganeti.

**Comment by apoikos** [comment]
This is certainly something that is useful, regardless of hotplugging support, and relatively easy to implement.

**Comment by anarcat** [comment]
> run_qemu_monitor_command: info block
> run_qemu_monitor_command: block_resize hotdisk-7d8a2ae9-pci-7 <any_size_smaller_then_current>G

i'm curious to hear more about this... everything i read online does this through `virsh` which, i suspect won't work with ganeti. what is that `run_qemu_monitor_command` command anyways?

**Comment by Ganneff** [comment]
Actually, resizing works nicely. Even online, and for /. That is, for
us, in our setup. :)

Needed: The gnt-instance grow-disk must work, without impact on the
VM. It does so here with extstorage, and that is what this is tested
on. I do not use drbd or plain or blockdev or whatnot.

Also, my system disks are set as one big partition.

But then it is:

`gnt-instance grow-disk to whatever new size`

Now on the host the VM runs on:

```
export instancename=the_full_name_of_the_instance_as_ganeti_knows_it
export diskno=0   # obviously the disk you just grew
export diskname=$(echo "info block" | socat STDIO UNIX-CONNECT:/var/run/ganeti/kvm-hypervisor/ctrl/$instancename.monitor|grep "$instancename:$diskno (raw)"|cut -d " " -f 1)
echo "block_resize $diskname 1" | /usr/bin/socat STDIO UNIX-CONNECT:/var/run/ganeti/kvm-hypervisor/ctrl/$instancename.monitor; echo
```

Thats it on the node. Now off into the VM.
Say, you resize the system disk and have the normal virtio in use, it will be /dev/vda

```
fdisk /dev/vda

p
```

Note the start and Id values, then delete the partition and recreate it.

```
d
n
```

Partition 1, primary, startsektor as noted, end one maximum, type same as before.
**DO NOT LET IT WIPE THE FILESYSTEM SIGNATURE, if it asks**.
Possibly set boot flag, if you had it, quit with writing (_w_)

Now, kick the system into knowing the new stuff:

```
partx -u /dev/vda1
resize2fs /dev/vda

```
Done.

**Comment by saz** [comment]
parted has a `resizepart` command, which is easier to use than fdisk (and maybe also less error prone.

I've not tried it, but there's a hook script by @saschalucas 
https://github.com/saschalucas/ganeti-hook-grow-disk

**Comment by Ganneff** [comment]
> parted has a `resizepart` command, which is easier to use than fdisk (and maybe also less error prone.

Not really, it appears it wants you to tell it the new length in sectors. fdisk does tell you that on its own.

> I've not tried it, but there's a hook script by @saschalucas

That looks interesting. I have been to lazy to morph mine into one, so maybe i can adjust and use this. Thanks.

**Comment by saz** [comment]
Works with a new size in MB/GB/TB or percent:

```
(parted) p
Model: Linux device-mapper (linear) (dm)
Disk /dev/dm-11: 524MB
Sector size (logical/physical): 512B/512B
Partition Table: gpt
Disk Flags: 

Number  Start   End     Size    File system  Name  Flags
 1      1049kB  99.6MB  98.6MB  ext4

(parted) resizepart                                                       
Partition number? 1                                                       
End?  [99.6MB]? 150m                                                      
(parted) p                                                                
Model: Linux device-mapper (linear) (dm)
Disk /dev/dm-11: 524MB
Sector size (logical/physical): 512B/512B
Partition Table: gpt
Disk Flags: 

Number  Start   End    Size   File system  Name  Flags
 1      1049kB  150MB  149MB  ext4

(parted) resizepart 1 100%
(parted) p                                                                
Model: Linux device-mapper (linear) (dm)
Disk /dev/dm-11: 524MB
Sector size (logical/physical): 512B/512B
Partition Table: gpt
Disk Flags: 

Number  Start   End    Size   File system  Name  Flags
 1      1049kB  524MB  523MB  ext4

(parted)  
```

**Comment by Ganneff** [comment]
Ah. Haha. I was looking at the (seperate) command resizepart, not inside parted. And that manpage has

> length The new length of the partition (in 512-byte sectors)

**Comment by theundefined** [comment]
I use "standarized" setup where all vm's have one pv /dev/vda1. And to grow i use something like:
gnt-instance grow-disk --absolute $name $disk $size (on ganeti master)
gnt-growonline $name (on ganeti primary node) - script https://github.com/theundefined/scripts/blob/master/gnt-growonline
growpart /dev/$diskname 1 (where diskname is usually vda) - inside vm

**Comment by r-tierney** [comment]
@theundefined your script works great, the only Modification i had to make was:
```
- blockname=$(grep $kvmname $tmpfile | grep -Eo '/var/run/ganeti/instance-disks/[a-z0-9:-]+')
+ blockname=$(grep $kvmname $tmpfile | awk '{print $3}')
```
As it didn't match my VM's hostname which was: `upgrade02-rya.ops` ( it would strip off the `-rya.ops` )

I ended up using the following Ganeti hook script as it allowed me to simply run `gnt-instance grow-disk upgrade02-rya.ops 0 2G` and it would automatically run the hook script to signal the update of the block size with no vm restart required.

https://github.com/saschalucas/ganeti-hook-grow-disk/blob/master/00-grow-disk-online
I made a slight modification to this one also PR here: https://github.com/saschalucas/ganeti-hook-grow-disk/pull/2

**Comment by codefritzel** [comment]
closed by #1731
